### ISSUE TITLE
Isn’t client ID and secret visible?

### ISSUE THREAD

**Issue by fakekamrankhan** [auth0/auth0-spa-js#408]
Hello,

I am opening this as a blank issue because I don’t think this is a bug or an improvement but just to get some clarification.

By using this approach, won’t your app be exposing the client ID and secret? Is this a known risk for using this approach or am I missing something here?

Any of the react static frameworks would change any environment variables during build time to their static values so I’m not sure how this risk is mitigated for single page applications?

**Comment by stevehobbsdev** [comment]
@fakekamrankhan Thanks for raising this!

There are a couple of things here. Firstly, the client ID itself is not considered a secret and is easily discoverable through a URL transaction anyway. It's simply an identifier for the IdP as to which application to use.

Secondly, you're correct in that the client secret _is_ a secret and that it is near impossible to store such secrets in an SPA. It's for this reason that such applications (and this SDK) should use the [Auth Code + PKCE](https://auth0.com/docs/flows/concepts/auth-code-pkce) flow, which is ideal for clients (browsers, native apps) where the Auth Code flow _without_ PKCE (which uses a client secret) could be compromised. Auth Code + PKCE does not require a client secret to be used, thus the client application has no need to store one. You can also [read more about Auth Code + PKCE](https://auth0.com/blog/oauth2-implicit-grant-and-spa/).

Finally, you're also correct in that it's good practise to configure the client ID (and client secret if not building an SPA) using the environment, but this is probably driven at least in part by a need to configure an app for different environments (local development vs. production, for example).

Hope that helps! Will close this for now but feel free to continue the discussion if you have questions.

**Comment by fakekamrankhan** [comment]
Thank you so much for the quick response!

My apologies, I just noticed that the client secret in this SDK is not required. 

If this SPA doesn’t require a client secret, am I to assume that by using the auth0-spa-js, my static react application is inherently using the Auth Code + PKCE flow if authentication is working as expected in the documentation here or is that something I need to double check within my application’s settings in Auth0?

**Comment by stevehobbsdev** [comment]
No problem!

> am I to assume that by using the auth0-spa-js, my static react application is inherently using the Auth Code + PKCE flow if authentication is working as expected

Yes! This SDK is designed to be very opinionated in that regard and does not allow you to use a different flow. Doing this (hopefully!) makes it easier for you to configure and use. If authentication is working for you, then rest assured you are already using the Auth Code + PKCE flow 👍

**Comment by victorkurauchi** [comment]
hi @stevehobbsdev thanks for your insights here.

I have a weird scenario, where I followed the React Quickstart https://auth0.com/docs/quickstart/spa/react#configure-auth0, and my Application type is set to use SPA and I configured the OIDC and JWT fields as in the article.

However, i still keep getting unauthorized usingg this lib. So if I go the logs I can see `Success Login` and then `Failed Exchange` right after.

Do you have any idea what I could be missing? I'm using the `auth0-react-01-login` repo.

What other way can I confirm that my Application is using Auth Code + PKCE flow ?

Thanks in advance!

**Comment by stevehobbsdev** [comment]
@victorkurauchi Which browser are you testing this flow in? Do you have anything that could be blocking third-party cookies (using Safari, Chrome with cookie-blocking extensions, etc).

To answer your question about PKCE, you can verify the flow by looking at the requests your browser network tab. PKCE requests will include a `code_challenge` param on the `/authorize` call, and a `code_verifier` param on the `/oauth/token` call. However, the mechanism we use by default to make the token exchange relies on your session cookies being available.

![image](https://user-images.githubusercontent.com/766403/79995048-9e1b8100-84ae-11ea-928a-a1df248aba2a.png)

**Comment by victorkurauchi** [comment]
Hi @stevehobbsdev ,

I tested in all browsers, Chrome Safari and Firefox.

One thing I noticed is that the /oauth/token post is failing. We are deploying the application via pipeline, so my config have this field:

`"token_endpoint_auth_method": "client_secret_post",`

it should probably have the value as `"token_endpoint_auth_method": "none",` because it's a SPA application thus we don't transport the client_secret? 

All the other configs are following the same pattern as in the Quickstart.

I'll try changing this value to `none` and see how it goes
Thanks for the reply
